engine/robot.py

```diff
@@ -330,8 +374,97 @@ def set_phase(self, new_phase):
+            Args:
+                dist_to_goal (Double): The distance from the robot to the goal at the start of the phase
+                prev_phase (Phase): The previous phase to return to after executing obstacle avoidance
+                init_threshold (Double): Radius from initial position that robot has to leave before it can detect if
+                    goal is unreachable
+                goal_threshold (Double): Threshold from goal that will be detected as reaching goal
+            Returns:
+                prev_phase (Phase)
+        """
+        init_x = self.x_pos
+        init_y = self.y_pos
+        # threshold is arbitrary, set later. sometimes location reading will be inaccurate or not frequent enough to
+        # read that we've arrived back to the init x y pos. make it small enough that robot actually leaves threshold
+        # at some time during boundary following or add timeout to branch making gate = True
+        gate = False
+        dist_to_goal_decreases = False
+        boundary_traversed = False
+        while True:
+            if not self.fault:  # fault has priority over obstacle avoidance
+                dist_from_init = math.sqrt((self.x_pos - init_x) ** 2 + (self.y_pos - init_y) ** 2)
+                if dist_from_init > init_threshold:
+                    gate = True
+                if curr_dist_to_goal < goal_threshold and gate:  # exits obstacle avoidance if robot close to goal
+                    self.set_phase(prev_phase)
+                    return None
+                elif boundary_traversed and gate:
+                    self.set_phase(Phase.FAULT)  # cannot reach goal
+                    return None
+                elif dist_to_goal_decreases:
+                    # don't include condition on whether there is a new obstacle bc it will be caught and
+                    # current algorithm will continue traversing current boundary (instead of traversing new obstacle)
+                    # not optimal because frequently will have to go back to obstacle avoidance
+                    self.set_phase(prev_phase)  # goes back to prev traversal
+                    return None
+                    # check position of goal and if obstacle in way of goal (using side sensors) then keep doing
+                    # boundary following except with new init_x, init_y, gate,
+                else:
+                    self.execute_boundary_following(0)  # add code directly here
+                    # update conditions
+                    curr_dist_to_goal = math.sqrt(
+                        (self.x_pos - self.goal_location[0]) ** 2 + (self.y_pos - self.goal_location[1]) ** 2)
+                    dist_to_goal_decreases = (curr_dist_to_goal < dist_to_goal)
+                    boundary_traversed = dist_from_init < init_threshold
+            else:
+                self.set_phase(Phase.FAULT)
+                return None
+            time.sleep(10)  # don't hog the cpu
```

Comment:
This makes sense in track_obstacle but why time sleep here

Reply:
so it allows the other (track_obstacle) thread to run. Otherwise, the cpu will never be allocated to the other thread... i think?

Reply:
ok ye makes sense! we can test out the sleep time later? i feel like 10 might be a bit long

Reply:
yeah, we can test it later